File: frontend/src/api/document.ts
import api from './client';
import type { Document } from '../types';

export interface CreateDocumentData {
  name: string;
  description?: string;
  category: string;
  url: string;
  file_type?: string;
  file_size?: number;
  document_date?: string;
  expiry_date?: string;
  tags?: string[];
  person_id?: string;
  property_id?: string;
  vehicle_id?: string;
  insurance_policy_id?: string;
  notes?: string;
}

export interface UpdateDocumentData extends Partial<CreateDocumentData> {}

export const documentApi = {
  // Get all documents for the user's household
  list: async (category?: string): Promise<Document[]> => {
    const params = category ? `?category=${category}` : '';
    const response = await api.get<Document[]>(`/documents${params}`);
    return response.data;
  },

  // Get a single document by ID
  get: async (id: string): Promise<Document> => {
    const response = await api.get<Document>(`/documents/${id}`);
    return response.data;
  },

  // Create a new document
  create: async (data: CreateDocumentData): Promise<Document> => {
    const response = await api.post<Document>('/documents', data);
    return response.data;
  },

  // Update a document
  update: async (id: string, data: UpdateDocumentData): Promise<Document> => {
    const response = await api.put<Document>(`/documents/${id}`, data);
    return response.data;
  },

  // Delete a document
  delete: async (id: string): Promise<void> => {
    await api.delete(`/documents/${id}`);
  },

  // Get expiring documents
  getExpiring: async (days: number = 30): Promise<Document[]> => {
    const response = await api.get<Document[]>(`/documents/expiring?days=${days}`);
    return response.data;
  },

  // Get documents by person
  getByPerson: async (personId: string): Promise<Document[]> => {
    const response = await api.get<Document[]>(`/documents/person/${personId}`);
    return response.data;
  },

  // Get category statistics
  getCategoryStats: async (): Promise<Record<string, number>> => {
    const response = await api.get<Record<string, number>>('/documents/stats');
    return response.data;
  },
};

File: api/internal/models/document.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Document categories
const (
	DocumentCategoryIdentity  = "IDENTITY"
	DocumentCategoryProperty  = "PROPERTY"
	DocumentCategoryVehicle   = "VEHICLE"
	DocumentCategoryInsurance = "INSURANCE"
	DocumentCategoryFinancial = "FINANCIAL"
	DocumentCategoryMedical   = "MEDICAL"
	DocumentCategoryLegal     = "LEGAL"
	DocumentCategoryOther     = "OTHER"
)

// Common file types
const (
	FileTypePDF  = "PDF"
	FileTypeDOC  = "DOC"
	FileTypeDOCX = "DOCX"
	FileTypeXLS  = "XLS"
	FileTypeXLSX = "XLSX"
	FileTypeJPG  = "JPG"
	FileTypePNG  = "PNG"
	FileTypeOther = "OTHER"
)

// Document represents a document link stored in the system
type Document struct {
	ID          uuid.UUID `json:"id" db:"id"`
	HouseholdID uuid.UUID `json:"household_id" db:"household_id"`

	// Document info
	Name        string  `json:"name" db:"name"`
	Description *string `json:"description,omitempty" db:"description"`
	Category    string  `json:"category" db:"category"`

	// Link to external storage
	URL      string  `json:"url" db:"url"`
	FileType *string `json:"file_type,omitempty" db:"file_type"`
	FileSize *int64  `json:"file_size,omitempty" db:"file_size"`

	// Dates
	DocumentDate *string `json:"document_date,omitempty" db:"document_date"`
	ExpiryDate   *string `json:"expiry_date,omitempty" db:"expiry_date"`

	// Tags
	Tags []string `json:"tags,omitempty" db:"tags"`

	// Linked entities
	PersonID          *uuid.UUID `json:"person_id,omitempty" db:"person_id"`
	PropertyID        *uuid.UUID `json:"property_id,omitempty" db:"property_id"`
	VehicleID         *uuid.UUID `json:"vehicle_id,omitempty" db:"vehicle_id"`
	InsurancePolicyID *uuid.UUID `json:"insurance_policy_id,omitempty" db:"insurance_policy_id"`

	Notes *string `json:"notes,omitempty" db:"notes"`

	CreatedAt time.Time `json:"created_at" db:"created_at"`
	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`

	// Computed fields
	DaysUntilExpiry *int  `json:"days_until_expiry,omitempty" db:"-"`
	IsExpired       *bool `json:"is_expired,omitempty" db:"-"`

	// Joined fields
	Person          *Person          `json:"person,omitempty" db:"-"`
	Property        *Property        `json:"property,omitempty" db:"-"`
	Vehicle         *Vehicle         `json:"vehicle,omitempty" db:"-"`
	InsurancePolicy *InsurancePolicy `json:"insurance_policy,omitempty" db:"-"`
}

// PopulateComputedFields calculates derived fields for the document
func (d *Document) PopulateComputedFields() {
	now := time.Now()

	// Calculate days until expiry
	if d.ExpiryDate != nil && *d.ExpiryDate != "" {
		if expiryDate, err := time.Parse("2006-01-02", *d.ExpiryDate); err == nil {
			days := int(expiryDate.Sub(now).Hours() / 24)
			d.DaysUntilExpiry = &days
			expired := expiryDate.Before(now)
			d.IsExpired = &expired
		}
	}
}

// ValidDocumentCategories returns valid document categories
func ValidDocumentCategories() []string {
	return []string{
		DocumentCategoryIdentity,
		DocumentCategoryProperty,
		DocumentCategoryVehicle,
		DocumentCategoryInsurance,
		DocumentCategoryFinancial,
		DocumentCategoryMedical,
		DocumentCategoryLegal,
		DocumentCategoryOther,
	}
}

// ValidFileTypes returns common file types
func ValidFileTypes() []string {
	return []string{
		FileTypePDF,
		FileTypeDOC,
		FileTypeDOCX,
		FileTypeXLS,
		FileTypeXLSX,
		FileTypeJPG,
		FileTypePNG,
		FileTypeOther,
	}
}
